File: javae/jdk/IceInternal/Connector.java
// **********************************************************************
//
// Copyright (c) 2003-2008 ZeroC, Inc. All rights reserved.
//
// This copy of Ice-E is licensed to you under the terms described in the
// ICEE_LICENSE file included in this distribution.
//
// **********************************************************************

package IceInternal;

final class Connector
{
    private static class ConnectThread extends Thread
    {
        ConnectThread(InetSocketAddress addr)
        {
            _addr = addr;
        }

        public void
        run()
        {
            try
            {
                java.net.Socket fd = new java.net.Socket(_addr.getAddress(), _addr.getPort());
                synchronized(this)
                {
                    _fd = fd;
                    notifyAll();
                }
            }
            catch(java.io.IOException ex)
            {
                synchronized(this)
                {
                    _ex = ex;
                    notifyAll();
                }
            }
        }

        java.net.Socket
        getFd(int timeout)
            throws java.io.IOException
        {
            java.net.Socket fd = null;

            synchronized(this)
            {
                while(_fd == null && _ex == null)
                {
                    try
                    {
                        wait(timeout);
                        break;
                    }
                    catch(InterruptedException ex)
                    {
                        continue;
                    }
                }

                if(_ex != null)
                {
                    throw _ex;
                }

                fd = _fd;
                _fd = null;
            }

            return fd;
        }

        private InetSocketAddress _addr;
        private java.net.Socket _fd;
        private java.io.IOException _ex;
    }

    public Transceiver
    connect(int timeout)
    {
        if(_traceLevels.network >= 2)
        {
            String s = "trying to establish tcp connection to " + toString();
            _logger.trace(_traceLevels.networkCat, s);
        }

        java.net.Socket fd = null;
        try
        {
            //
            // If a connect timeout is specified, do the connect in a separate thread.
            //
            if(timeout >= 0)
            {
                ConnectThread ct = new ConnectThread(_addr);
                ct.start();
                fd = ct.getFd(timeout == 0 ? 1 : timeout);
                if(fd == null)
                {
                    throw new Ice.ConnectTimeoutException();
                }
            }
            else
            {
                fd = new java.net.Socket(_addr.getAddress(), _addr.getPort());
            }
            Network.setTcpBufSize(fd, _instance.initializationData().properties, _logger);
        }
        catch(java.net.ConnectException ex)
        {
            if(fd != null)
            {
                try
                {
                    fd.close();
                }
                catch(java.io.IOException e)
                {
                }
            }
            Ice.ConnectFailedException se;
            if(Network.connectionRefused(ex))
            {
                se = new Ice.ConnectionRefusedException();
            }
            else
            {
                se = new Ice.ConnectFailedException();
            }
            se.initCause(ex);
            throw se;
        }
        catch(java.io.IOException ex)
        {
            if(fd != null)
            {
                try
                {
                    fd.close();
                }
                catch(java.io.IOException e)
                {
                }
            }
            Ice.SocketException e = new Ice.SocketException();
            e.initCause(ex);
            throw e;
        }
        catch(RuntimeException ex)
        {
            if(fd != null)
            {
                try
                {
                    fd.close();
                }
                catch(java.io.IOException e)
                {
                }
            }
            throw ex;
        }

        if(_traceLevels.network >= 1)
        {
            String s = "tcp connection established\n" + IceInternal.Network.fdToString(fd);
            _logger.trace(_traceLevels.networkCat, s);
        }

        return new Transceiver(_instance, fd);
    }

    public String
    toString()
    {
        return Network.addrToString(_addr);
    }

    //
    // Only for use by TcpEndpoint
    //
    Connector(Instance instance, InetSocketAddress addr, int timeout)
    {
        _instance = instance;
        _addr = addr;
        _timeout = timeout;
        _traceLevels = instance.traceLevels();
        _logger = instance.initializationData().logger;

    }

    //
    // Compare connectors for sorting purposes
    //
    public boolean
    equals(java.lang.Object obj)
    {
        return compareTo(obj) == 0;
    }

    public int
    compareTo(java.lang.Object obj) // From java.lang.Comparable
    {
        Connector p = null;

        try
        {
            p = (Connector)obj;
        }
        catch(ClassCastException ex)
        {
            return 1;
        }

        if(this == p)
        {
            return 0;
        }

        if(_timeout < p._timeout)
        {
            return -1;
        }
        else if(p._timeout < _timeout)
        {
            return 1;
        }

        return Network.compareAddress(_addr, p._addr);
    }

    private Instance _instance;
    private InetSocketAddress _addr;
    private int _timeout;
    private TraceLevels _traceLevels;
    private Ice.Logger _logger;
}

File: objc/src/Transports/Connector.cpp
// **********************************************************************
//
// Copyright (c) 2003-2008 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICETOUCH_LICENSE file included in this distribution.
//
// **********************************************************************

#include <Transceiver.h>
#include <EndpointI.h>
#include <Connector.h>

#include <Ice/Instance.h>
#include <Ice/TraceLevels.h>
#include <Ice/LoggerUtil.h>
#include <Ice/Network.h>
#include <Ice/Exception.h>

#include <CoreFoundation/CoreFoundation.h>

using namespace std;
using namespace Ice;
using namespace IceInternal;

TransceiverPtr
IceObjC::Connector::connect()
{
    if(_traceLevels->network >= 2)
    {
        Trace out(_logger, _traceLevels->networkCat);
        out << "trying to establish " << _instance->protocol() << " connection to " << toString();
    }

    try
    {
        //
        // Create the read/write streams
        //
        CFReadStreamRef readStream;
        CFWriteStreamRef writeStream;
        CFDataRef addr = nil;
        if(_addr.ss_family == AF_INET)
        {
            addr = CFDataCreate(NULL, reinterpret_cast<const UInt8*>(&_addr), sizeof(sockaddr_in));
        }
        else if(_addr.ss_family == AF_INET6)
        {
            addr = CFDataCreate(NULL, reinterpret_cast<const UInt8*>(&_addr), sizeof(sockaddr_in6));
        }
        CFHostRef host = CFHostCreateWithAddress(NULL, addr);
        CFRelease(addr);
        CFStreamCreatePairWithSocketToCFHost(NULL, host, getPort(_addr), &readStream, &writeStream);
        CFRelease(host);

        _instance->setupStreams(readStream, writeStream, false, _host);

        SOCKET fd;
        CFDataRef d = (CFDataRef)CFReadStreamCopyProperty(readStream, kCFStreamPropertySocketNativeHandle);
        CFDataGetBytes(d, CFRangeMake(0, sizeof(SOCKET)), reinterpret_cast<UInt8*>(&fd));
        CFRelease(d);

        if(!CFReadStreamSetProperty(readStream, kCFStreamPropertyShouldCloseNativeSocket, kCFBooleanFalse) || 
           !CFWriteStreamSetProperty(writeStream, kCFStreamPropertyShouldCloseNativeSocket, kCFBooleanFalse))
        {
            throw Ice::ConnectFailedException(__FILE__, __LINE__, 0);
        }

        setBlock(fd, false);
        setTcpBufSize(fd, _instance->initializationData().properties, _logger);

        return new Transceiver(_instance, fd, readStream, writeStream, false, _host);
    }
    catch(const Ice::LocalException& ex)
    {
        if(_traceLevels->network >= 2)
        {
            Trace out(_logger, _traceLevels->networkCat);
            out << "failed to establish " << _instance->protocol() << " connection to ";
            out << toString() << "\n" << ex;
        }
        throw;
    }
}

Short
IceObjC::Connector::type() const
{
    return _instance->type();
}

string
IceObjC::Connector::toString() const
{
    return addrToString(_addr);
}

bool
IceObjC::Connector::operator==(const IceInternal::Connector& r) const
{
    const Connector* p = dynamic_cast<const Connector*>(&r);
    if(!p)
    {
        return false;
    }

    if(compareAddress(_addr, p->_addr) != 0)
    {
        return false;
    }

    if(_timeout != p->_timeout)
    {
        return false;
    }

    if(_connectionId != p->_connectionId)
    {
        return false;
    }

    return true;
}

bool
IceObjC::Connector::operator!=(const IceInternal::Connector& r) const
{
    return !operator==(r);
}

bool
IceObjC::Connector::operator<(const IceInternal::Connector& r) const
{
    const Connector* p = dynamic_cast<const Connector*>(&r);
    if(!p)
    {
        return type() < r.type();
    }

    if(_timeout < p->_timeout)
    {
        return true;
    }
    else if(p->_timeout < _timeout)
    {
        return false;
    }

    if(_connectionId < p->_connectionId)
    {
        return true;
    }
    else if(p->_connectionId < _connectionId)
    {
        return false;
    }
    return compareAddress(_addr, p->_addr) == -1;
}

IceObjC::Connector::Connector(const InstancePtr& instance, 
                              const struct sockaddr_storage& addr, 
                              Ice::Int timeout, 
                              const string& connectionId,
                              const string& host) :
    _instance(instance),
    _traceLevels(instance->traceLevels()),
    _logger(instance->initializationData().logger),
    _addr(addr),
    _timeout(timeout),
    _connectionId(connectionId),
    _host(host)
{
}

IceObjC::Connector::~Connector()
{
}
